jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdCredentials.java

```diff
@@ -0,0 +1,201 @@
+        Object azp = claims.get("azp");
+        if (azp != null && !configuration.getClientId().equals(azp))
+        {
+            LOG.warn("Authorized party claim value should be the client_id");
+            throw new IllegalArgumentException("invalid azp claim");
+        }
+    }
+
+    public boolean isExpired()
+    {
+        if (authCode != null || claims == null)
+            return true;
+
+        // Check expiry
+        long expiry = (Long)claims.get("exp");
+        long currentTimeSeconds = (long)(System.currentTimeMillis() / 1000F);
+        if (currentTimeSeconds > expiry)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("OpenId Credentials expired {}", this);
+            return true;
+        }
+
+        return false;
+    }
+
+    protected Map<String, Object> decodeJWT(String jwt) throws IOException
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("decodeJWT {}", jwt);
+
+        String[] sections = jwt.split("\\.");
+        if (sections.length != 3)
+            throw new IllegalArgumentException("JWT does not contain 3 sections");
+
+        String jwtHeaderString = new String(Base64.getDecoder().decode(sections[0]), StandardCharsets.UTF_8);
+        String jwtClaimString = new String(Base64.getDecoder().decode(sections[1]), StandardCharsets.UTF_8);
+        String jwtSignature = sections[2];
+
+        Map<String, Object> jwtHeader = (Map)JSON.parse(jwtHeaderString);
+        LOG.debug("JWT Header: {}", jwtHeader);
+
+        /* If the ID Token is received via direct communication between the Client
+         and the Token Endpoint (which it is in this flow), the TLS server validation
+          MAY be used to validate the issuer in place of checking the token signature. */
+        if (LOG.isDebugEnabled())
+            LOG.debug("JWT signature not validated {}", jwtSignature);
+
+        return (Map)JSON.parse(jwtClaimString);
+    }
+
+    private Map<String, Object> claimAuthCode(String authCode) throws IOException
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("claimAuthCode {}", authCode);
+
+        // Use the authorization code to get the id_token from the OpenID Provider
+        String urlParameters = "code=" + authCode +
+            "&client_id=" + configuration.getClientId() +
+            "&client_secret=" + configuration.getClientSecret() +
+            "&redirect_uri=" + redirectUri +
+            "&grant_type=authorization_code";
+
+        URL url = new URL(configuration.getTokenEndpoint());
+        HttpURLConnection connection = (HttpURLConnection)url.openConnection();
+        connection.setDoOutput(true);
+        connection.setRequestMethod("POST");
+        connection.setRequestProperty("Host", configuration.getOpenIdProvider());
+        connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+        connection.setRequestProperty("charset", "utf-8");
```

Comment:
I don't think this is doing what you think.
`setRequestProperty` adds a HTTP header, so you really need a `charset` header, or you just wanted to specify the charset on the `Content-Type` on the line above?